internal/internal_workflow_client.go

```diff
@@ -877,6 +927,69 @@ func (wc *WorkflowClient) ResetWorkflowExecution(ctx context.Context, request *w
 	return resp, nil
 }
 
+func (wc *WorkflowClient) UpdateWorkflowWithOptions(
+	ctx context.Context,
+	req *UpdateWorkflowWithOptionsRequest,
+) (WorkflowUpdateHandle, error) {
+	if err := wc.ensureInitialized(); err != nil {
+		return nil, err
+	}
+	argPayloads, err := wc.dataConverter.ToPayloads(req.Args...)
+	if err != nil {
+		return nil, err
+	}
+	header, _ := headerPropagated(ctx, wc.contextPropagators)
+
+	grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+	defer cancel()
+	resp, err := wc.workflowService.UpdateWorkflow(grpcCtx, &workflowservice.UpdateWorkflowRequest{
+		RequestId: uuid.New(),
+
+		// currently fixed as this is the only supported execution method - will
+		// change to BlockingPolicy (or similar) when upstream API changes land
+		ResultAccessStyle: enumspb.WORKFLOW_UPDATE_RESULT_ACCESS_STYLE_REQUIRE_INLINE,
+		Namespace:         wc.namespace,
+		WorkflowExecution: &commonpb.WorkflowExecution{
+			WorkflowId: req.WorkflowID,
+			RunId:      req.RunID,
+		},
+		Identity:            wc.identity,
+		FirstExecutionRunId: req.FirstExecutionRunID,
+		Input: &interactionpb.Input{
+			Header: header,
+			Name:   req.UpdateName,
+			Args:   argPayloads,
+		},
+	})
+	if err != nil {
+		return &updateHandle{value: err}, nil
+	}
+	if failure := resp.Output.GetFailure(); failure != nil {
+		return &updateHandle{value: wc.failureConverter.FailureToError(failure)}, nil
```

Comment:
For queries, on the server side we differentiate all other errors from user-defined worker errors with a "query failed error" gRPC status details. See https://github.com/temporalio/api-go/blob/8a0bd676b97b61e2f8bd254389a3c9b96ed278b3/serviceerror/convert.go#L114. It seems here we are intentionally not reusing gRPC errors for update failed errors, which is fine, but I think we need to make sure somewhere that the caller can differentiate from update failure and all other server failures. For queries today they can check if it's `serviceerror.QueryFailed`. We need something similar for updates too (even if the error is defined SDK side instead of API side).

Reply:
Not there yet with the server implementation but I take your point. :+1:

Reply:
Is the server impl expected to set unique error details that represent worker-side update failure? If so and we add a new type to https://pkg.go.dev/go.temporal.io/api/serviceerror to support that, no prob. But then why even have a `Output.GetFailure()` instead of just failing the gRPC call if you're always gonna fail the gRPC call in a specific way in this case? I don't mind either way and other SDKs will be making their own `WorkflowUpdateFailed` errors that differentiate.

Reply:
But this isn't a gRPC failure, so the server side won't help you completely. You're not gonna add a new type in https://pkg.go.dev/go.temporal.io/api/serviceerror because that's only for gRPC failures right? Since this is not a gRPC failure, are you expecting to create a update-failure type in https://github.com/temporalio/api/blob/master/temporal/api/failure/v1/message.proto? Whether you do or not, https://pkg.go.dev/go.temporal.io/sdk/temporal will need a `WorkflowUpdateError` type.

Basically the caller of this call should know whether it was a worker-side update failure just like the caller of queries know it's a worker-side query failure. I don't mind how we implement it, but if you keep it as not a gRPC error, you're gonna need a type in this SDK repo to represent it.